### DISCUSSION TITLE
HTML Minifier?

### DISCUSSION THREAD

**Discussion by ivodolenc** [swc-project/swc#6278] [Q&A]
Hi, I just found out about this project through Vercel's new blog [post](https://nextjs.org/blog/next-13). I had a quick look through the documentation so I hope this question isn't ridiculous :D

Is it possible to minify the `html` content? Maybe something like [this](https://swc.rs/docs/configuration/minification#swcminifycode-options), but for `html`:

```js
// Example

import { readFileSync, writeFileSync } from 'fs'
import swc from '@swc/core'

let file = 'src/index.html'

let htmlContent = readFileSync(file, 'utf-8')

let minifiedContent = await swc.minify(htmlContent, {
  // options...
})

writeFileSync(file, minifiedContent)
```

**Comment by ivodolenc** [comment]
Hi, thanks for comment.

Sorry if I didn't express myself well, I was thinking if this is possible with `swc` out of the box, since `Next.js 13` will be powered with `swc` under the hood? 😊

In the new Next.js docs, minification is mentioned as one of the new big features, and also migration from terser.

Nothing about this topic was mentioned in the official `swc` docs (at least for JavaScript API), but in some `PR`s and issues it was noted that html minification is supported now so I was hoping I could get more info on that.

**Reply by kdy1** [reply]
It's supported, but it's barely documented
@alexander-akait Do you know the way to use `@swc/html` at now?

**Reply by alexander-akait** [reply]
@ivodolenc Hello, yes, here example of usage https://github.com/webpack-contrib/html-minimizer-webpack-plugin/blob/master/src/utils.js#L112, two function:
- minify (when you have full HTML document, i.e. with Doctype/head/body)
- minifyFragment (i.e. part of HTML, i.e. no Doctype/etc, like you have HTML template in separate file)

Why two? Because spec separates how to parser them. You need to install `@swc/html`, in future maybe we revisit it and move them to `@swc/code`, now we keep them separately. 

Here repo with `@swc/html` and `@swc/css` - https://github.com/swc-project/bindings, feel free to feedback

**Comment by ivodolenc** [comment]
@alexander-akait Hi, thanks for all the info and examples, it definitely helped.

Here's an example with which I was partially able to set up the minification, but it seems to automatically remove the `<head>`, `<body>` and the final `</html>` tag.

The rest of the content is minified and works fine in the browser, but is not correct without the removed tags.

Although I think I have set something wrong, here is my example, maybe you have some advice:

```js
// vite-plugin for SvelteKit

import swcHtml from '@swc/html'

{
  return {
    name: 'vite-plugin',

    async closeBundle() {
      // htmlFiles -> array of all .html files in dir

      for (let file of htmlFiles) {
        // open file and get full html content with doctype and everything else
        let fileContent = readFileSync(file, 'utf-8')

        // minify html content
        // not sure about that Buffer.from(), is that set correctly?
        let minifiedContent = await swcHtml.minify(Buffer.from(fileContent))

        // write minified content back to the file
        writeFileSync(file, minifiedContent.code)
      }
    }
  }
}
```

Also, can't believe that `@swc/html` is only `~7kb` min?? Game-changer 🤯

**Reply by alexander-akait** [reply]
> Here's an example with which I was partially able to set up the minification, but it seems to automatically remove the <head>, <body> and the final </html> tag.

Because you don't need it for HTML 5 docs by spec :smile: Try it in all HTML5 browsers, anyway if you need to support very old browser (do you really need IE6-7 :smile: ?) You can pass the `tag_omission: false` option https://github.com/swc-project/bindings/blob/main/crates/html_node/src/lib.rs#L154

You can read about tag omission more in spec (i.e. no `</li>` ends, no `<tbody></tbody>` and more, I think you heard about it)

Also please use `Promise.all` and wait all promises otherwise you can have some problems with non finished async tasks

**Reply by ivodolenc** [reply]
> Because you don't need it for HTML 5 docs by spec 😄 Try it in all HTML5 browsers, anyway if you need to support very old browser (do you really need IE6-7 😄 ?)

Honestly, this is the first time I've heard that I don't need those tags. All websites still use them that way, so it doesn't get any attention at all. 

Ok, so everything works as it should. Thanks again for your help.

**Reply by alexander-akait** [reply]
@ivodolenc 

> Honestly, this is the first time I've heard that I don't need those tags. All websites still use them that way, so it doesn't get any attention at all.

Yeah, maybe because we don't have full spec compliance HTML parser and minifier before in tooling (js solutions mostly based on regexps)

Ref: https://html.spec.whatwg.org/multipage/sections.html#the-body-element

> [Tag omission in text/html](https://html.spec.whatwg.org/multipage/dom.html#concept-element-tag-omission):
A [body](https://html.spec.whatwg.org/multipage/sections.html#the-body-element) element's [start tag](https://html.spec.whatwg.org/multipage/syntax.html#syntax-start-tag) can be omitted if the element is empty, or if the first thing inside the [body](https://html.spec.whatwg.org/multipage/sections.html#the-body-element) element is not [ASCII whitespace](https://infra.spec.whatwg.org/#ascii-whitespace) or a [comment](https://html.spec.whatwg.org/multipage/syntax.html#syntax-comments), except if the first thing inside the [body](https://html.spec.whatwg.org/multipage/sections.html#the-body-element) element is a [meta](https://html.spec.whatwg.org/multipage/semantics.html#the-meta-element), [noscript](https://html.spec.whatwg.org/multipage/scripting.html#the-noscript-element), [link](https://html.spec.whatwg.org/multipage/semantics.html#the-link-element), [script](https://html.spec.whatwg.org/multipage/scripting.html#the-script-element), [style](https://html.spec.whatwg.org/multipage/semantics.html#the-style-element), or [template](https://html.spec.whatwg.org/multipage/scripting.html#the-template-element) element.
A [body](https://html.spec.whatwg.org/multipage/sections.html#the-body-element) element's [end tag](https://html.spec.whatwg.org/multipage/syntax.html#syntax-end-tag) can be omitted if the [body](https://html.spec.whatwg.org/multipage/sections.html#the-body-element) element is not immediately followed by a [comment](https://html.spec.whatwg.org/multipage/syntax.html#syntax-comments).
